backend/fbdev.c

```diff
@@ -43,22 +57,43 @@ typedef struct {
     size_t fb_len;
 } twin_fbdev_t;
 
-static void _twin_fbdev_put_span(twin_coord_t left,
-                                 twin_coord_t top,
-                                 twin_coord_t right,
-                                 twin_argb32_t *pixels,
-                                 void *closure)
+static void _twin_fbdev_put_span16(twin_coord_t left,
+                                   twin_coord_t top,
+                                   twin_coord_t right,
+                                   twin_argb32_t *pixels,
+                                   void *closure)
 {
-    twin_screen_t *screen = SCREEN(closure);
-    twin_fbdev_t *tx = PRIV(closure);
+    uint32_t *dest;
+    GET_TWIN_FBDEV(left, top, closure, &dest);
+    twin_coord_t width = right - left;
+    for (int i = 0; i < width; i++) {
+        dest[i] = ARGB32_TO_RGB565(pixels[i]);
+    }
+}
 
-    if (tx->fb_base == MAP_FAILED)
-        return;
+static void _twin_fbdev_put_span24(twin_coord_t left,
+                                   twin_coord_t top,
+                                   twin_coord_t right,
+                                   twin_argb32_t *pixels,
+                                   void *closure)
+{
+    uint32_t *dest;
+    GET_TWIN_FBDEV(left, top, closure, &dest);
+    twin_coord_t width = right - left;
+    for (int i = 0; i < width; i++) {
+        dest[i] = ARGB32_TO_RGB888(pixels[i]);
+    }
+}
 
+static void _twin_fbdev_put_span32(twin_coord_t left,
+                                   twin_coord_t top,
+                                   twin_coord_t right,
+                                   twin_argb32_t *pixels,
+                                   void *closure)
+{
+    uint32_t *dest;
+    GET_TWIN_FBDEV(left, top, closure, &dest);
     twin_coord_t width = right - left;
-    off_t off = top * screen->width + left;
-    uint32_t *dest =
-        (uint32_t *) ((uintptr_t) tx->fb_base + (off * sizeof(*dest)));
     memcpy(dest, pixels, width * sizeof(*dest));
 }
```

Comment:
Use macros to generate the similar functions. Check `MEM_READ_IMPL` and `MEM_WRITE_IMPL` in file [src/io.c](https://github.com/sysprog21/rv32emu/blob/master/src/io.c) of rv32emu.

Reply:
Got it.